### DISCUSSION TITLE
Use rebase instead of merge method to update PRs when using queue action

### DISCUSSION THREAD

**Discussion by mzwaen** [Mergifyio/mergify#2849] [Q&A]
Hi! we are new to mergify and would like to use this tool to solve this specific situation:

We would like PRs to be automatically merged to the develop branch (our default branch) when certain conditions are met, in particular, when several CI checks (including GitHub actions) passed.
PRs that are out of date should be rebased to develop, CI tests should then be retriggered and the PRs should only be merged if the CI checks were successful.

I saw that the `strict_method=rebase` option should be able to achieve this, and I was going to use this configuration that seems to work well for  the use case described above:
```
   actions:
      merge:
        strict: smart
        strict_method: rebase
        method: merge
```

but the documentation says that this option is now depreciated: https://docs.mergify.io/actions/merge/

It suggests replacing the `merge` action with the `queue` action and I therefore tried this configuration:

```
    actions:
      queue:
        name: default
        method: merge
```

However, from what I can see, using the `queue` action, there is no option to _update_ the PRs using the `rebase` method rather than the `merge` method. With the `merge` method, the new changes in develop are merged into the PR branch and mergify adds a merge commit to it, which is not what we want.

Does a `rebase` method exist for the update of PRs when using the `queue` action? I know that the `rebase` method has some drawbacks:
https://docs.mergify.io/actions/merge/#strict-rebase
but we would be ready to live with them.

If the option to use the `rebase` method is not available for the `queue` action, are there risks in continuing using the depreciated `Strict merge` option? According to the documentation, it should be able to prevent merging broken pull requests as it should retrigger all CI tests after the PR has been updated. Could you confirm this is the case and we can keep using this depreciated method (so the first configuration)?

Thank you very much in advance!

**Comment by jd** [comment]
Hi @mzwaen,

There's no `rebase` update method available for the `queue` action, mostly because when using `speculative_checks`, only merge operations are supported to test the code speculatively. Therefore we did not (re)implement the `rebase` update method to keep things simple.

I'm surprised that you're using the `merge` method if you're using `rebase` to update the PR. In general, we've seen people more interested into having a linear history using `rebase` as the merge method.

What's the underlying reading for wanting `rebase` as the update strategy? What'd be wrong for you with a regular `merge`?

**Reply by mzwaen** [reply]
Hi! Thanks for your quick answer! The main reason why we would like to keep `rebase` as the update strategy is to keep our history consistent: until now we always rebased (manually) our PR branches when they were out of date before merging them to develop, which means we only have merge commits when merging feature branches to develop (and not the other way round) which we find allows for a very clean (and easy to understand) history. This makes it easy to search in the history for merge commits and know that it always corresponds to a point in which a feature branch was merged into develop.
Would you be considering adding `rebase` as the update strategy for the `queue` action in the future? If not, could you tell us what the drawbacks of using this configuration are? 
```
   actions:
      merge:
        strict: smart
        strict_method: rebase
        method: merge
```
In particular, can we still safely assume that this configuration will not allow broken pull requests to be merged (for exemple because the CI tests would be running on an outdated branch?)

**Reply by jd** [reply]
Thanks for the explanation @mzwaen. That makes sense.

We'll see into adding the rebase method for updating PR with the `queue` action.

The only downside to using the configuration you are proposing is that you can't use speculative checks.

Anyhow, we will be phasing `strict` mode soon, but in the meantime, it'll work. By then we should get the rebase method implemented in the `queue` action so you'll be able to transition easily. :)

**Reply by mzwaen** [reply]
This is great! Thanks a lot @jd for your explanations and for planning to add the feature to the `queue` action, this will be very helpful for us!

**Reply by jd** [reply]
@mzwaen We've added `rebase` as a method to the queue action.

**Reply by AraHaan** [reply]
@jd Can it be possible to have the update method of the queue to be both: rebase+squash (mergify will only use the squash part of it if pr's commits is > 1) while wanting to merge the pr with only rebase?

Currently my configuration uses:
```yml
      merge:
        strict: smart+fasttrack
        strict_method: rebase
        method: squash
```
On one of the rules to some of the repositories that I maintain.

**Reply by jd** [reply]
I'm not sure what the point of squashing after a rebase would be? You'd lost history while trying to update a PR, that sounds like a bad idea.
Could you elaborate?

**Reply by AraHaan** [reply]
Sure, basically on my PR's I want them to be rebase merged to a single commit after it uses rebase to update followed by a squash if the branches commits are 2 ahead of the base branch to make it 1 ahead of it and then it pushes it to the PR branch where then the CI reruns and if it passes it gets merged.

Alternatively it *could* be made to squash first then rebase on top of the base branch when it updates the PR. Either way works for me.
